File: Napakalaki/lib/cultist_player.rb
# encoding: utf-8

require_relative 'player'

module NapakalakiGame
  
  class CultistPlayer < Player
    @@totalCultistPlayers = 0

    def initialize(p, c)
      copy_const(p)
      @myCultistCard = c
      
      @@totalCultistPlayers += 1
    end

    protected

    # El nivel de combate de los jugadores sectarios se obtiene con la fórmula siguiente
    def getCombatLevel
      (1.7 * super + @myCultistCard.getGainedLevels * @@totalCultistPlayers).to_i
    end

    # Algunos monstruos cambian el nivel de combate contra sectarios
    def getOponentLevel(m)
      m.getCombatLevelAgainstCultistPlayer
    end

    def shouldConvert
      false
    end

    def giveMeATreasure
      @visibleTreasures.delete_at(rand(@visibleTreasures.length))
    end

    def canYouGiveMeATreasure
      not @visibleTreasures.empty?
    end

    public

    def self.getTotalCultistPlayers
      @@totalCultistPlayers
    end
    
    def to_s
      super + "\n - Carta de sectario: " + @myCultistCard.to_s
    end

  end # CultistPlayer
  
end # NapakalakiGame

File: Napakalaki/src/NapakalakiGame/CultistPlayer.java
/**
 * CultistPlayer.java
 */

package NapakalakiGame;

import java.util.Random;

/**
 * Representa un jugador sectario 
 */
public class CultistPlayer extends Player {
    
    private static int totalCultistPlayers = 0;
    private Cultist myCultistCard;
    
    public CultistPlayer(Player p, Cultist c) {
        super(p);
        myCultistCard = c;
        totalCultistPlayers++;
    }
    
    @Override
    protected int getCombatLevel() {
        // super.getCombatLevel() + 0.7 * super.getCombatLevel() = 1.7 * super.getCombatLevel()
        return (int) (1.7 * super.getCombatLevel()) + 
                      myCultistCard.getGainedLevels() * totalCultistPlayers;
    }
    
    @Override
    protected int getOponentLevel(Monster m) {
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    @Override
    protected boolean shouldConvert() {
        return false;
    }
    
    @Override
    protected boolean canYouGiveMeATreasure() {
        return !getVisibleTreasures().isEmpty();
    }
    
    @Override
    protected Treasure giveMeATreasure() {
        return getVisibleTreasures().remove(new Random().nextInt(getVisibleTreasures().size()));
    }
    
    public static int getTotalCultistPlayers() {
        return totalCultistPlayers;
    }
    
    @Override
    public String toString() {
        return super.toString() + "\n - Carta de sectario: " + myCultistCard.toString();
    }
    
}
